### PR TITLE
ARROW-15172: [Go] Add Arm64 Neon implementation for Arrow-math

### PR DESCRIPTION
**Pull request by guyuqi**
Add 'sum_int64_neon', 'sum_uint64_neon', 'sum_floatt64_neon'  with Arm64 Golang assembly. C2GOASM doesn't work correctly for Arm64.
`uint64_neon_arm64.s`  `int64_neon_arm64.s` `float64_neon_arm64.s` were partly generated by asm2plan9s.

-  Add ability to enable Arm64 extension via environment:

   `ARM_ENABLE_EXT=NEON`

- Benchmark
Disable Arm64 Neon:
```
goos: linux
goarch: arm64
pkg: github.com/apache/arrow/go/v7/arrow/math
BenchmarkFloat64Funcs_Sum_256-46                 3247237               364.9 ns/op      5611.91 MB/s
BenchmarkFloat64Funcs_Sum_1024-46                1000000              1424 ns/op        5750.99 MB/s
BenchmarkFloat64Funcs_Sum_8192-46                 100620             11360 ns/op        5768.89 MB/s
BenchmarkFloat64Funcs_Sum_1000000-46                 866           1382561 ns/op        5786.36 MB/s
BenchmarkInt64Funcs_Sum_256-46                   5655325               251.3 ns/op      8150.04 MB/s
BenchmarkInt64Funcs_Sum_1024-46                  1254841               954.0 ns/op      8586.80 MB/s
BenchmarkInt64Funcs_Sum_8192-46                   148898              7515 ns/op        8720.33 MB/s
BenchmarkInt64Funcs_Sum_1000000-46                  1299            921258 ns/op        8683.78 MB/s
BenchmarkUint64Funcs_Sum_256-46                  4753304               246.4 ns/op      8313.19 MB/s
BenchmarkUint64Funcs_Sum_1024-46                 1253706               954.7 ns/op      8580.65 MB/s
BenchmarkUint64Funcs_Sum_8192-46                  149168              7561 ns/op        8667.80 MB/s
BenchmarkUint64Funcs_Sum_1000000-46                 1304            918844 ns/op        8706.60 MB/s
```
Enable Arm64 Neon:
```
goos: linux
goarch: arm64
pkg: github.com/apache/arrow/go/v7/arrow/math
BenchmarkFloat64Funcs_Sum_256-46                11145474               102.4 ns/op      19996.84 MB/s
BenchmarkFloat64Funcs_Sum_1024-46                3156472               375.5 ns/op      21816.63 MB/s
BenchmarkFloat64Funcs_Sum_8192-46                 351138              2886 ns/op        22707.53 MB/s
BenchmarkFloat64Funcs_Sum_1000000-46                3456            346282 ns/op        23102.55 MB/s
BenchmarkInt64Funcs_Sum_256-46                  11427655               101.4 ns/op      20196.81 MB/s
BenchmarkInt64Funcs_Sum_1024-46                  3231861               373.0 ns/op      21963.24 MB/s
BenchmarkInt64Funcs_Sum_8192-46                   382744              2880 ns/op        22753.01 MB/s
BenchmarkInt64Funcs_Sum_1000000-46                  3486            344900 ns/op        23195.16 MB/s
BenchmarkUint64Funcs_Sum_256-46                 11319964               100.9 ns/op      20306.60 MB/s
BenchmarkUint64Funcs_Sum_1024-46                 3180728               373.8 ns/op      21914.31 MB/s
BenchmarkUint64Funcs_Sum_8192-46                  368254              2881 ns/op        22748.54 MB/s
BenchmarkUint64Funcs_Sum_1000000-46                 3481            345534 ns/op        23152.55 MB/s
```

Get 2.5X performance uplift for Int64, 2.6X for Uint64 and 3.8X for Float64.

### CODE REVIEW

**Comment by guyuqi** [comment]
Test results:
```
$ uname -a
5.8.0-48-generic #54~20.04.1-Ubuntu SMP Sat Mar 20 13:31:53 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux

$ make ARM_ENABLE_EXT=NEON test
go test  ./...
ok      github.com/apache/arrow/go/v7/arrow     (cached)
ok      github.com/apache/arrow/go/v7/arrow/array       (cached)
ok      github.com/apache/arrow/go/v7/arrow/arrio       (cached)
ok      github.com/apache/arrow/go/v7/arrow/bitutil     (cached)
?       github.com/apache/arrow/go/v7/arrow/cdata       [no test files]
ok      github.com/apache/arrow/go/v7/arrow/csv (cached)
ok      github.com/apache/arrow/go/v7/arrow/decimal128  (cached)
?       github.com/apache/arrow/go/v7/arrow/endian      [no test files]
ok      github.com/apache/arrow/go/v7/arrow/flight      (cached)
ok      github.com/apache/arrow/go/v7/arrow/float16     (cached)
?       github.com/apache/arrow/go/v7/arrow/internal/arrdata    [no test files]
ok      github.com/apache/arrow/go/v7/arrow/internal/arrjson    (cached)
ok      github.com/apache/arrow/go/v7/arrow/internal/cpu        (cached)
?       github.com/apache/arrow/go/v7/arrow/internal/debug      [no test files]
?       github.com/apache/arrow/go/v7/arrow/internal/flatbuf    [no test files]
?       github.com/apache/arrow/go/v7/arrow/internal/flight_integration [no test files]
?       github.com/apache/arrow/go/v7/arrow/internal/flight_integration/cmd/arrow-flight-integration-client     [no test files]
?       github.com/apache/arrow/go/v7/arrow/internal/flight_integration/cmd/arrow-flight-integration-server     [no test files]
?       github.com/apache/arrow/go/v7/arrow/internal/testing/gen        [no test files]
ok      github.com/apache/arrow/go/v7/arrow/internal/testing/tools      (cached)
?       github.com/apache/arrow/go/v7/arrow/internal/testing/types      [no test files]
ok      github.com/apache/arrow/go/v7/arrow/ipc (cached)
ok      github.com/apache/arrow/go/v7/arrow/ipc/cmd/arrow-cat   (cached)
ok      github.com/apache/arrow/go/v7/arrow/ipc/cmd/arrow-file-to-stream        (cached)
ok      github.com/apache/arrow/go/v7/arrow/ipc/cmd/arrow-json-integration-test (cached)
ok      github.com/apache/arrow/go/v7/arrow/ipc/cmd/arrow-ls    (cached)
ok      github.com/apache/arrow/go/v7/arrow/ipc/cmd/arrow-stream-to-file        (cached)
ok      github.com/apache/arrow/go/v7/arrow/math        (cached)
ok      github.com/apache/arrow/go/v7/arrow/memory      (cached)
ok      github.com/apache/arrow/go/v7/arrow/scalar      (cached)
ok      github.com/apache/arrow/go/v7/arrow/tensor      (cached)

```

**Review by cyb70289 — APPROVED** [review]
> Get 2.5X performance uplift for Int64, 2.6X for Uint64 and 3.8X for Float64.

Great work! Thanks @guyuqi !

File: go/arrow/math/Makefile [hunk 1]
```diff
@@ -37,27 +41,42 @@ INTEL_SOURCES := \
...
+	uint64_neon_arm64.s
+
 .PHONEY: assembly
 
-assembly: $(INTEL_SOURCES)
+#ifeq ($(GOARCH), arm64)
+assembly: $(ARM_SOURCES)
+#else
+#assembly: $(INTEL_SOURCES)
+#endif
```
Comment (zeroshade) [inline go/arrow/math/Makefile:56]: Does this only work if you're currently *on* arm64? I believe Go has arguments and env vars which allow for creating / building cross platform, is it possible to build the ARM sources for Arm64 neon without having to be on an arm64 neon machine? Either way, can we have comments or the readme updated with whatever instructions / requirements are needed to generate the arm64 sources?

**Review by zeroshade — CHANGES_REQUESTED** [review]
Thanks so much @guyuqi this is awesome. Just had one comment, but otherwise this is great.

File: go/arrow/math/Makefile (hunk 1 above)
Comment (guyuqi) [reply go/arrow/math/Makefile]: @zeroshade Thanks for your comments. :-)
Agreed on what you mentioned. 

But after further consideration, it's noted that `c2goasm`doesn't work correctly for Arm64. 
The Arm64 assembly were generated by `asm2plan9s`, and I manually formatted the related assembler codes.

IMO, it seems that we would not add the `assembly` target for Arm64 until `c2goasm` added the Arm64 support.  
The `ARM_SOURCES` should be removed.

Comment (zeroshade) [reply go/arrow/math/Makefile]: @guyuqi Sounds good to me, can we just document either in a comment or in the readme the procedure for re-generating the arm64 assembly including a comment about the fact that c2goasm doesn't support it? I'm happy once there's a comment or documentation in the readme just so we don't lose this info if we lose track of this PR.

Comment (guyuqi) [reply go/arrow/math/Makefile]: Agreed. I'll add a comment in assembly files to retain the related info to track this PR, thanks.

Comment (guyuqi) [reply go/arrow/math/Makefile]: Updated.

**Review by zeroshade — APPROVED** [review]

**Comment by zeroshade** [comment]
Thanks @guyuqi!

**Pull request closed without merging**